src/CodeGen_Renderscript_Dev.cpp

```diff
@@ -94,66 +94,70 @@ void CodeGen_Renderscript_Dev::add_kernel(Stmt stmt, const std::string &kernel_n
-                                   GlobalValue::CommonLinkage,
-                                   0,  // has initializer, specified below
-                                   arg_name);
+            enum RS_ARGUMENT_TYPE { RS_INT = 6, RS_BUFFER = 20 } rs_argument_type;
+            GlobalVariable *gvar;
+            if (!args[i].is_buffer) {
+                gvar = new GlobalVariable(*module, llvm::Type::getInt32Ty(*context),
+                                          false,  // isConstant
+                                          GlobalValue::CommonLinkage,
+                                          0,  // has initializer, specified below
+                                         arg_name);
+                gvar->setInitializer(const_0);
+                globals_sym_names.push_back(std::make_tuple(arg_name, gvar));
+
+                rs_argument_type = RS_INT;
+            } else {
+                gvar = new GlobalVariable(*module, StructTy_struct_rs_allocation,
+                                          false,  // isConstant
+                                          GlobalValue::CommonLinkage,
+                                          0,  // has initializer, specified below
+                                          arg_name);
+                gvar->setInitializer(const_empty_allocation_struct);
+
+                rs_argument_type = RS_BUFFER;
+                rs_objects_slots->addOperand(
+                    MDNode::get(*context,
+                                vec<LLVMMDNodeArgumentType>(MDString::get(
+                                    *context,
+                                    std::to_string(rs_export_var->getNumOperands())))));
+            }
             gvar->setAlignment(4);
             globals_sym_names.push_back(std::make_tuple(arg_name, gvar));
-            gvar->setInitializer(const_empty_allocation_struct);
+            rs_global_vars.insert(std::pair<std::string,GlobalVariable*>(arg_name, gvar));
 
-            //
-            //  "20" means "buffer"-type for Renderscript.
-            //
             rs_export_var->addOperand(MDNode::get(
                 *context,
                 vec<LLVMMDNodeArgumentType>(MDString::get(*context, arg_name),
-                                            MDString::get(*context, "20"))));
-            rs_objects_slots->addOperand(
-                MDNode::get(*context, vec<LLVMMDNodeArgumentType>(MDString::get(
-                                          *context, std::to_string(i)))));
+                                            MDString::get(*context, std::to_string(rs_argument_type)))));
+
+            Halide::Type type = args[i].type;
+            debug(2) << "args[" << i << "] = {"
+                     << "name=" << args[i].name
+                     << " is_buffer=" << args[i].is_buffer
+                     << " dimensions=" << args[i].dimensions << " type.code="
+                     << (type.code == Halide::Type::TypeCode::Int
```

Comment:
If you include IRPrinter.h you can pass a Halide::Type directly here instead of unpacking it.